### PR TITLE
Adds support for Actor Reentrancy

### PR DESCRIPTION
**Pull request by berndverst**
# Description
Adds support for Actor Reentrancy

## Issue reference
https://github.com/dapr/python-sdk/issues/214
https://github.com/dapr/dapr/issues/2901

Please reference the issue this PR will close: #214

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [X] Code compiles correctly
* [X] Created/updated tests
* [ ] Extended the documentation

### CODE REVIEW

**Comment by berndverst** [comment]
Hold off on reviewing. Will add some tests. Also it appears an existing GRPC test (which should be unrelated) is failing.

**Comment by wcs1only** [comment]
> Hold off on reviewing. Will add some tests. Also it appears an existing GRPC test (which should be unrelated) is failing.

Forgive me if you know this already: You can actually denote a PR as a 'draft' to indicate that it isn't ready to be reviewed/merged. When you go to create a new PR, there's a little dropdown on the right side of the submit button that will let you 'Create as draft'

**Comment by berndverst** [comment]
@wcs1only I think this is ready for review now.

By the way, how exactly are the samples run? Do you think I should add a sample to this PR or do that in a separate PR?

File: dapr/clients/http/dapr_actor_http_client.py
Comment (berndverst) [inline dapr/clients/http/dapr_actor_http_client.py:88]: There was an unused variable.

**Comment by berndverst** [comment]
@halspang can you verify this looks good as far as the actors functionality goes (the right things are sent in the right spots)

File: dapr/actor/runtime/config.py
Comment (halspang) [inline dapr/actor/runtime/config.py:15]: While the reentrancy config itself is optional, this isn't an optional field in the config. How does this affect the actual serialization?

File: dapr/clients/http/dapr_actor_http_client.py
Comment (halspang) [inline dapr/clients/http/dapr_actor_http_client.py:53]: Can you explain how this makes a circular dependency?

File: dapr/clients/http/dapr_actor_http_client.py
Comment (halspang) [inline dapr/clients/http/dapr_actor_http_client.py:56]: Can this erase other headers that were already set/needed? Or is this the only header? I'm thinking of things like content-type or length.

File: tests/actor/test_actor_reentrancy.py
Comment (halspang) [inline tests/actor/test_actor_reentrancy.py:48]: To confirm, this is just validating that we can call a reentrant actor normally right? Not actually testing the reentrant portion?

File: tests/actor/test_actor_reentrancy.py
Comment (halspang) [inline tests/actor/test_actor_reentrancy.py:77]: Should this be the passthrough method? That's the only time the reentrant ID is fetched from the ContextVar.

**Review by halspang — COMMENTED** [review]
Overall looks good. Just a few comments. For the header/reentrancy itself I think we're good aside from the config parsing.

File: dapr/clients/http/dapr_actor_http_client.py
Comment (berndverst) [reply dapr/clients/http/dapr_actor_http_client.py:53]: I honestly can't. When the import was at the module / library level all kinds of tests failed because the wrong things got imported. Doing this import here fixed the issue. The same thing is done in other places in the Python SDK. Basically what is described here: https://stackabuse.com/python-circular-imports

File: dapr/actor/runtime/config.py
Comment (berndverst) [reply dapr/actor/runtime/config.py:15]: This does the right thing because of the default value, but the Optional qualifier is unnecessary. I'll remove this.

File: tests/actor/test_actor_reentrancy.py
Comment (berndverst) [reply tests/actor/test_actor_reentrancy.py:48]: This simulates that a reentrant can be called normally way (as the final call in a reentrancy chain for example) but it only tests the internal part where the HTTP library has already parsed the reentrancy header and then calls the runtime dispatch function shown here. This particular test doesn't test calling a reentrant actor from a reentrant actor. That's done in another test.

File: tests/actor/test_actor_reentrancy.py
Comment (berndverst) [reply tests/actor/test_actor_reentrancy.py:77]: I don't understand this comment. Could you explain @halspang?

This is just a bunch of Python weirdness to run two async methods truly in parallel and gather the results to force the interleaving of the processing. It is a sanity check to make sure the ContextVars truly work as we expected, that is, ContextVars are unique to each AsyncTask and are always copied forward into nested AsyncTasks.

I understand this looks weird - @wcs1only and I paired on this.

File: dapr/clients/http/dapr_actor_http_client.py
Comment (berndverst) [reply dapr/clients/http/dapr_actor_http_client.py:56]: Previously no custom headers were sent at all. That hasn't been supported with Actors in Python. The `invoke_method` only supports a single payload that is byte-encoded. This also only returns bytes.

The behavior prior to this PR is as follows:
Invoke_method always sends an empty dictionary for the headers via `send_bytes()`. Then that function checks whether the Content Type and DAPR API Token headers are presented (which they aren't) and if not uses defaults. See the `send_bytes` method definition and the test modification I made.

This introduces no breaking change.

**Comment by berndverst** [comment]
Thanks for the review @halspang. Made a small change (removing the Optional qualifier which wasn't effective anyway). Otherwise take a look at my reply to your comments. Hope it all makes sense :)

File: tests/actor/test_actor_reentrancy.py
Comment (halspang) [reply tests/actor/test_actor_reentrancy.py:77]: I'm fine with the python itself, my comment was more on how we'd handle the ContextVar in the code. Right now, this path only exercises the setting of the reentrant ID. If we use the passthrough method, it should also have to fetch the reentrant ID from the ContextVar. That's the path I'd like to check with interleaving.

Comment (halspang) [reply tests/actor/test_actor_reentrancy.py:77]: @berndverst what do you think about my previous comment?

**Review by halspang — APPROVED** [review]
Spoke over a call about my last comment and I think what I'm looking for is more of an IT or auto-example. Once we have a good idea on what to make the actor do, we can add that in but we don't need to in this PR.

File: dapr/clients/http/dapr_actor_http_client.py
Comment (wcs1only) [reply dapr/clients/http/dapr_actor_http_client.py:53]: IIRC, the actual circular dependency was on dapr.actor.runtime which in turn depends on the http client.

**Review by wcs1only — APPROVED** [review]
Looks good. +1 for adding an example that exercises this.

**Pull request merged**
